src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -768,25 +779,48 @@ private void WriteHeaderCollection(HttpHeaders headers)
                 if (knownHeader != null)
                 {
                     // The Host header is not sent for HTTP2 because we send the ":authority" pseudo-header instead
-                    // (see pseudo-header handling below in WriteHeaders). The Connection header is also not supported in HTTP2.
-                    if (knownHeader != KnownHeaders.Host && knownHeader != KnownHeaders.Connection)
+                    // (see pseudo-header handling below in WriteHeaders).
+                    // The Connection, Upgrade and ProxyConnection headers are also not supported in HTTP2.
+                    if (knownHeader != KnownHeaders.Host && knownHeader != KnownHeaders.Connection && knownHeader != KnownHeaders.Upgrade && knownHeader != KnownHeaders.ProxyConnection)
                     {
+                        if (header.Key.KnownHeader == KnownHeaders.TE)
+                        {
+                            // HTTP/2 allows only 'trailers' TE header. rfc7540 8.1.2.2
+                            foreach (string value  in header.Value)
+                            {
+                                if (value == "trailers")
```

Comment:
This is doing a case-sensitive comparison; does it need to be ordinal case insensitive?

Reply:
I could not find definitive answer. This is not header name so general requirement for case-insensitive comparison does not apply. The spec says:

>   The only exception to this is the TE header field, which MAY be
>    present in an HTTP/2 request; when it is, it MUST NOT contain any
>    value other than "trailers".
>  

Since I could not find any reference suggesting otherwise, I assume this is exact match. 
I don't know if using StringComparison.InvariantCulture would be more appropriate.

Reply:
I know that we've recieved bug reports where users ran into issues with a similar assumption about header value casing. Unless this is a paticularly hot path, I think it's probably worth making the comparision case insensitive.

Reply:
I thought the HTTP/2 protocol always transmitted headers as lowercase.  So, shouldn't we assume they already are?

Reply:
this is header value. (TE: trailers)